worker/raft/raftbackstop/worker.go

```diff
@@ -0,0 +1,257 @@
+}
+
+// This worker monitors the state of the raft cluster it's passed, and
+// if it detects that there is only one remaining API server that is
+// non-voting, will force-append a configuration message to the raft
+// log to enable it to recover. This allows us to handle the situation
+// where we had two remaining machines (a leader and a non-voting
+// follower) but the leader was removed.
+
+// Config holds the values needed by the worker.
+type Config struct {
+	Raft     RaftNode
+	LogStore raft.LogStore
+	Hub      *pubsub.StructuredHub
+	Logger   loggo.Logger
+	Tag      names.Tag
+}
+
+// Validate validates the raft worker configuration.
+func (config Config) Validate() error {
+	if config.Hub == nil {
+		return errors.NotValidf("nil Hub")
+	}
+	if config.Raft == nil {
+		return errors.NotValidf("nil Raft")
+	}
+	if config.LogStore == nil {
+		return errors.NotValidf("nil LogStore")
+	}
+	if config.Tag == nil {
+		return errors.NotValidf("nil Tag")
+	}
+	return nil
+}
+
+// NewWorker returns a worker responsible for recovering the raft
+// cluster when it's been reduced to one server that can't become
+// leader.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := &Worker{
+		config:        config,
+		serverDetails: make(chan apiserver.Details),
+	}
+	// Subscribe to API server address changes.
+	unsubscribe, err := config.Hub.Subscribe(
+		apiserver.DetailsTopic,
+		w.apiserverDetailsChanged,
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "subscribing to apiserver details")
+	}
+	// Now that we're subscribed, request the current API server details.
+	req := apiserver.DetailsRequest{
+		Requester: "raft-backstop",
+		LocalOnly: true,
+	}
+	if _, err := config.Hub.Publish(apiserver.DetailsRequestTopic, req); err != nil {
+		return nil, errors.Annotate(err, "requesting current apiserver details")
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			defer unsubscribe()
+			return w.loop()
+		},
+	}); err != nil {
+		unsubscribe()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+type Worker struct {
+	catacomb catacomb.Catacomb
+	config   Config
+
+	serverDetails chan apiserver.Details
+	configUpdated bool
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *Worker) loop() error {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case details := <-w.serverDetails:
+			err := w.maybeRecoverCluster(details)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (w *Worker) maybeRecoverCluster(details apiserver.Details) error {
+	if w.configUpdated {
+		w.config.Logger.Infof("raft configuration already updated; waiting for raft worker restart")
```

Comment:
How many times do we expect a serverDetails channel event? We could get many of these log message spewed out? Should we have a flag so that only one is printed?

Reply:
Yeah, good point - I'll add the flag.